libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -1051,11 +1051,11 @@ void CompilerUtils::convertType(
 	case Type::Category::ArraySlice:
 	{
 		auto& typeOnStack = dynamic_cast<ArraySliceType const&>(_typeOnStack);
-		solUnimplementedAssert(
-			_targetType.dataStoredIn(DataLocation::CallData),
-			"Conversion from calldata slices to memory not yet implemented."
-		);
-		solAssert(_targetType == typeOnStack.arrayType(), "");
+		solAssert(_targetType.category() == Type::Category::Array, "");
+		auto const& targetArrayType = dynamic_cast<ArrayType const&>(_targetType);
+		solAssert(targetArrayType == *typeOnStack.arrayType().copyForLocation(targetArrayType.location(), false), "");
```

Comment:
I think this condition is more restrictive than the one in `ArraySliceType::isImplicitlyConvertibleTo()` so there will be some cases that pass through type checker just fine and fail here.

Looking at [`ArrayType::isImplicitlyConvertibleTo()`](https://github.com/ethereum/solidity/blob/v0.7.5/libsolidity/ast/Types.cpp#L1684-L1722) I think that for example `int[] calldata` -> `uint[] storage` should be one such types.

Reply:
I have changed assertions in both places to be the same, and to align to `TypeChecker`. Will add more tests now, for converting of base types that are not the same.